clash-cores/src/Clash/Cores/LineCoding8b10b.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE CPP #-}
+
+-- |
+--   Copyright   :  (C) 2024, QBayLogic B.V.
+--   License     :  BSD2 (see the file LICENSE)
+--   Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+--
+--   8b/10b encoding and decoding functions
+module Clash.Cores.LineCoding8b10b where
+
+import qualified Clash.Cores.LineCoding8b10b.Decoder as Dec
+import qualified Clash.Cores.LineCoding8b10b.Encoder as Enc
+import Clash.Prelude
+
+-- | Data type that contains a 'BitVector' with the corresponding error
+--   condition of the decode function
+data Symbol8b10b
+  = -- | Correct data word
+    Dw (BitVector 8)
+  | -- | Correct control word
+    Cw (BitVector 8)
+  | -- | Incorrect data word
+    DwError (BitVector 8)
+  | -- | Running disparity error
+    RdError (BitVector 8)
+  deriving (Generic, NFDataX, Eq, Show)
+
+-- | Function to check whether a 'Symbol8b10b' results in a data word
+isDw :: Symbol8b10b -> Bool
+isDw (Dw _) = True
+isDw _ = False
+
+-- | Function to check whether a 'Symbol8b10b' is not an error value
+isValidSymbol :: Symbol8b10b -> Bool
+isValidSymbol (Cw _) = True
+isValidSymbol dw = isDw dw
+
+-- | Function to convert a 'Symbol8b10b' to a plain 'BitVector 8'
+fromDw :: Symbol8b10b -> BitVector 8
+fromDw dw = case dw of
+  Dw _dw -> _dw
+  Cw _dw -> _dw
+  _ -> 0
+
+-- | Take the running disparity and the current code group, and return a tuple
+--   containing the new running disparity and a 'Symbol8b10b' containing the
+--   decoded value. This function uses a 'MemBlob' to store the decoder lookup
+--   table.
+decode8b10b ::
+  -- | Running disparity
+  Bool ->
+  -- | Code group
+  BitVector 10 ->
+  -- | Tuple containing the new running disparity and the 'Symbol8b10b'
+  (Bool, Symbol8b10b)
+decode8b10b rd cg = (rdNew, dw)
+ where
+  dw
+    | cgEr = DwError (pack _dw)
+    | rdEr = RdError (pack _dw)
+    | cw = Cw (pack _dw)
+    | otherwise = Dw (pack _dw)
+
+  (statusBits, _dw) =
+    splitAt d4
+      $ bv2v
+      $ asyncRomBlobPow2
+        $(memBlobTH Nothing Dec.decoderLut)
+      $ unpack (bitCoerce rd ++# cg)
+
+  rdEr = bitCoerce $ head statusBits
+  cgEr = bitCoerce $ statusBits !! (1 :: Index 4)
+  cw = bitCoerce $ statusBits !! (2 :: Index 4)
+  rdNew = bitCoerce $ last statusBits
```

Comment:
```suggestion
decode8b10b rd cg = (rdNew, sym)
 where
  sym
    | cgEr = DwError w
    | rdEr = RdError w
    | cw = Cw w
    | otherwise = Dw w

  (rdEr, cgEr, cw, rdNew, w) =
    unpack
      $ asyncRomBlobPow2
        $(memBlobTH Nothing Dec.decoderLut)
      $ unpack (pack rd ++# cg)
```

So `MemBlob` loses type information; the only thing it remembers about the type of the data elements is their bit size. But you would usually just `unpack` it back from the data output of the ROM. In this case, since the data inside the `MemBlob` has the type `(Bit, Bit, Bit, Bit, BitVector 8)`, you would usually `unpack` it back into that type. However, I presume that  `Bit` is used here only for the ease of the copy-paste from the source. They are in fact all four `Bool`s, and from a Haskell standpoint, it would have made more sense to write the table with the type `(Bool, Bool, Bool, Bool, BitVector 8)`. So that's what we'll `unpack` it to here. We don't even need to specify that type; it follows from how the resulting bindings are used.

(From a Haskell standpoint, a direct encoding as `(Bool, Symbol8b10b)` would have made even more sense, but  that is even further from the copy-paste).

And on a totally different note, I think your use of the term _code group_ is wrong here. I'm pretty sure that a 10-bit entity is not a code group, but that it consists of two code groups, the first being 5b6b and the second being 3b4b.

We already use the word _symbol_ for things of type `Symbol8b10b`, so let's not call the 10-bit quantity that as well. Let's call it a _code_. I quickly scoured some standards and it seems a not uncommon term. Likewise, I think that the 5b6b section is a _code_ and the 3b4b section is a _code_, and that those together form two _code groups_. So the "group" isn't inherent to the parts, it's because they are grouped together in the 10-bit quantity.

Reply:
Also reasonable, thank you!

As for the naming, IEEE 802.3 consistently calls these 10-bit things a "code group", so I took that. I'd personally prefer to stick to that.

Reply:
Right, 802.3, the 5600-page tome I should have checked instead of elsewhere.

Reply:
7025-page 😉

Reply:
We should compare copies :-P. Mine is 802.3-2018. I got it legally and everything. It's exactly 5600 pages, nice round number.

Reply:
Hahaha ah yes, mine is 802.3-2022. Downloaded in the first week of my internship, also legal as I can just turn on EduVPN and click "Download".